### PR TITLE
Add readable prompts for permission claims

### PR DESCRIPTION
**Pull request by nkkmpf**
- Add human readable prompts for permission claims to cli plugin 
- Add unit tests for permission claim prompts
- Fixed small typo

TODOs:

- [x] Rebase code into functions
- [x] Add unit tests for individual flags and name them properly
- [x] Cover the case that no Selector or Owner is set

### CODE REVIEW

File: pkg/kubectl/bind-apiservice/plugin/servicebindings_test.go [hunk 1]
```diff
@@ -0,0 +1,756 @@
...
+
+func TestHumanReadablePromt(t *testing.T) {
+	t.Parallel()
+
+	tests := []struct {
+		name           string
+		testData       kubebindv1alpha1.PermissionClaim
+		expectedOutput string
+	}{
+		{"Owner=Provider",
```
Comment (lhaendler) [inline pkg/kubectl/bind-apiservice/plugin/servicebindings_test.go:37]: **issue** Missing test cases for donate and adopt without selector e.g.
```go
		{"Selector={},donate=true",
			kubebindv1alpha1.PermissionClaim{
				GroupResource: kubebindv1alpha1.GroupResource{
					Group:    "",
					Resource: "foo",
				},
				Version:  "v1",
				Selector: kubebindv1alpha1.ResourceSelector{},
				Required: true,
				Create: &kubebindv1alpha1.CreateOptions{
					Donate: true,
				},
			},
			"Create and modify user-owned foo objects (apiVersion: \"v1\") on your cluster.\n" +
				"The provider will not be able to modify or delete said objects.\n" +
				"Accepting this Permission is required in order to proceed.\n" +
				"Do you accept this Permission? [No,Yes]\n",
		},
		{"Selector={},adopt=true,update.preserving=[\"spec\"]",
			kubebindv1alpha1.PermissionClaim{
				GroupResource: kubebindv1alpha1.GroupResource{
					Group:    "",
					Resource: "foo",
				},
				Version:  "v1",
				Selector: kubebindv1alpha1.ResourceSelector{},
				Adopt:    true,
				Update: &kubebindv1alpha1.UpdateOptions{
					Preserving: []string{"spec"},
				},
			},
			"The provider wants to become owner of all user-created foo objects (apiVersion: \"v1\").\n" +
				"The provider will be able to access, modify and delete said objects on your cluster.\n" +
				"The following fields of the objects will still be managed by the user:\n" +
				"\t\"spec\"\n" +
				"Accepting this Permission is optional.\n" +
				"Do you accept this Permission? [No,Yes]\n",
		},
```

File: pkg/kubectl/bind-apiservice/plugin/servicebindings.go [hunk 2]
```diff
@@ -139,13 +141,182 @@ func (b *BindAPIServiceOptions) createAPIServiceBindings(ctx context.Context, co
...
+
+	donate := false
+	if claim.Create != nil {
+		donate = claim.Create.Donate
+	}
+
+	owner := claim.Selector.Owner
+	switch {
+	case (owner == kubebindv1alpha1.Provider && !donate && claim.Update.Fields != nil):
+		fallthrough
```
Comment (lhaendler) [inline pkg/kubectl/bind-apiservice/plugin/servicebindings.go:267]: The use of fallthrough makes this hard to read. Have you considered combining these cases into one conditional?

File: pkg/kubectl/bind-apiservice/plugin/servicebindings.go [hunk 3]
```diff
@@ -139,13 +141,182 @@ func (b *BindAPIServiceOptions) createAPIServiceBindings(ctx context.Context, co
...
+	case (owner == kubebindv1alpha1.Provider && donate && claim.Update.Preserving != nil):
```
Comment (lhaendler) [inline pkg/kubectl/bind-apiservice/plugin/servicebindings.go:268]: These expressions have a lot of repetition (e.g. repeating owner checks). Have you tried simplifying them? It could make sense to introduce helper variables with semantic meanings.

File: pkg/kubectl/bind-apiservice/plugin/servicebindings_test.go [hunk 4]
```diff
@@ -0,0 +1,756 @@
+/*
+Copyright 2022 The Kube Bind Authors.
```
Comment (lhaendler) [inline pkg/kubectl/bind-apiservice/plugin/servicebindings_test.go:2]:
```suggestion
Copyright 2023 The Kube Bind Authors.
```

File: pkg/kubectl/bind-apiservice/plugin/servicebindings.go [hunk 5]
```diff
@@ -139,13 +141,182 @@ func (b *BindAPIServiceOptions) createAPIServiceBindings(ctx context.Context, co
 	return bindings, nil
 }
 
+//Print Permission Claim to provided Writer
```
Comment (lhaendler) [inline pkg/kubectl/bind-apiservice/plugin/servicebindings.go:144]: **suggestion** This comment provides little additional information over the function signature. Could you expand or remove it?

File: pkg/kubectl/bind-apiservice/plugin/servicebindings_test.go [hunk 6]
```diff
@@ -0,0 +1,756 @@
...
+			},
+			"Create and modify the following provider owned foo objects (apiVersion: \"v1\") on your cluster referenced with:\n" +
+				"	name: \"bar\"\n" +
+				"Accepting this Permission is required in order to proceed.\n" +
+				"Do you accept this Permission? [No,Yes]\n",
+		},
+		{"Owner=Provider,GroupResource.Group",
+			kubebindv1alpha1.PermissionClaim{
+				GroupResource: kubebindv1alpha1.GroupResource{
+					Group:    "bar",
```
Comment (lhaendler) [inline pkg/kubectl/bind-apiservice/plugin/servicebindings_test.go:90]: **suggestion**```suggestion
					Group:    "example.com",
```

File: pkg/kubectl/bind-apiservice/plugin/servicebindings_test.go [hunk 7]
```diff
@@ -0,0 +1,756 @@
...
+					Resource: "foo",
+				},
+				Version: "v1",
+				Selector: kubebindv1alpha1.ResourceSelector{
+					Owner: kubebindv1alpha1.Provider,
+				},
+				Required: true,
+			},
+			"Create and modify provider owned foo objects (apiVersion: \"bar/v1\") on your cluster.\n" +
```
Comment (lhaendler) [inline pkg/kubectl/bind-apiservice/plugin/servicebindings_test.go:99]: **note**: the object is provider owned because create.donate is not specified. Not because the selector is owner=provider. The selector owner=provider just means that the reconciler will only look at provider owned objects (if a consumer owned object exists, the reconciler will skip it)

File: pkg/kubectl/bind-apiservice/plugin/servicebindings.go [hunk 8]
```diff
@@ -139,13 +141,182 @@ func (b *BindAPIServiceOptions) createAPIServiceBindings(ctx context.Context, co
...
+	var err error
+
+	donate := false
+	if claim.Create != nil {
+		donate = claim.Create.Donate
+	}
+
+	selectorName := claim.Selector.Name
+
+	switch {
```
Comment (lhaendler) [inline pkg/kubectl/bind-apiservice/plugin/servicebindings.go:202]: This string building can be split up to reduce duplication.
For example 
```go
fmt.Fprintf(b, "Create and modify %s on your cluster", <"user-owned"/"provider owned">)
fmt.Fprintf(b, "referenced with: %s", ...)
fmt.Fprintf(b, "\n")
```

**Review by lhaendler — CHANGES_REQUESTED** [review]
Provider and user owned objects use a different sentence structure, One starts with "The provider", the other does not.
The selector is only used for excluding objects from reconciliation, splitting the semantics on it may lead to confusion, and missing some use cases as shown by my suggested test cases.

**Comment by a9s-niklas-kempf** [comment]
> Provider and user owned objects use a different sentence structure, One starts with "The provider", the other does not. The selector is only used for excluding objects from reconciliation, splitting the semantics on it may lead to confusion, and missing some use cases as shown by my suggested test cases.

It seems that I misunderstood the role of the selector and the ownership of the claim (Adopt&Donate). I hope I now understood it as it was intended and corrected the code & tests accordingly.

File: pkg/kubectl/bind-apiservice/plugin/servicebindings_test.go (hunk 1 above)
Comment (a9s-niklas-kempf) [reply pkg/kubectl/bind-apiservice/plugin/servicebindings_test.go:37]: I added (more) test cases with an empty selector in 926bb5993775516f6dbfe28787228a9b11d95f92. I think i should now cover all cases.

File: pkg/kubectl/bind-apiservice/plugin/servicebindings.go (hunk 2 above)
Comment (a9s-niklas-kempf) [reply pkg/kubectl/bind-apiservice/plugin/servicebindings.go]: Thats true, initially I thought that I would end up with more different cases. I changed it to conditionals in 926bb5993775516f6dbfe28787228a9b11d95f92.

File: pkg/kubectl/bind-apiservice/plugin/servicebindings.go (hunk 3 above)
Comment (a9s-niklas-kempf) [reply pkg/kubectl/bind-apiservice/plugin/servicebindings.go]: I reworked and simplified this in 926bb5993775516f6dbfe28787228a9b11d95f92.

File: pkg/kubectl/bind-apiservice/plugin/servicebindings_test.go (hunk 4 above)
Comment (a9s-niklas-kempf) [reply pkg/kubectl/bind-apiservice/plugin/servicebindings_test.go:2]: Corrected this in 926bb5993775516f6dbfe28787228a9b11d95f92.

File: pkg/kubectl/bind-apiservice/plugin/servicebindings_test.go (hunk 6 above)
Comment (a9s-niklas-kempf) [reply pkg/kubectl/bind-apiservice/plugin/servicebindings_test.go]: Changed in 926bb5993775516f6dbfe28787228a9b11d95f92

File: pkg/kubectl/bind-apiservice/plugin/servicebindings_test.go (hunk 7 above)
Comment (a9s-niklas-kempf) [reply pkg/kubectl/bind-apiservice/plugin/servicebindings_test.go]: Reworked this in 926bb5993775516f6dbfe28787228a9b11d95f92.

File: pkg/kubectl/bind-apiservice/plugin/servicebindings.go (hunk 8 above)
Comment (a9s-niklas-kempf) [reply pkg/kubectl/bind-apiservice/plugin/servicebindings.go:190]: I simplified this in 926bb5993775516f6dbfe28787228a9b11d95f92

File: pkg/kubectl/bind-apiservice/plugin/servicebindings.go [hunk 9]
```diff
@@ -139,13 +141,156 @@ func (b *BindAPIServiceOptions) createAPIServiceBindings(ctx context.Context, co
...
+	if (claim.Selector != kubebindv1alpha1.ResourceSelector{}) {
+		name = claim.Selector.Name
+		owner = claim.Selector.Owner
+	}
+	var postamble string
+	switch {
+	case !donate && !adopt:
+		groupResource = "read " + groupResource
+	case donate && !adopt:
+		groupResource = "create " + groupResource
```
Comment (lhaendler) [inline pkg/kubectl/bind-apiservice/plugin/servicebindings.go:194]: Donate means the ownership is assigned to the consumer

File: pkg/kubectl/bind-apiservice/plugin/servicebindings.go [hunk 10]
```diff
@@ -139,13 +141,156 @@ func (b *BindAPIServiceOptions) createAPIServiceBindings(ctx context.Context, co
...
+	case !donate && adopt:
+		groupResource = "have ownership of " + groupResource
+		postamble = " and assign the ownership to you"
+	case donate && adopt:
+		groupResource = "create" + groupResource
+		postamble = " and wants to keep the ownership"
+	}
+
+	_, err = fmt.Fprintf(b, "The provider wants to %s on your cluster%s.", groupResource, postamble) //xxx
```
Comment (lhaendler) [inline pkg/kubectl/bind-apiservice/plugin/servicebindings.go:203]: What does xxx mean?

File: pkg/kubectl/bind-apiservice/plugin/servicebindings_test.go [hunk 11]
```diff
@@ -0,0 +1,807 @@
...
+				Selector: kubebindv1alpha1.ResourceSelector{},
+				Create: &kubebindv1alpha1.CreateOptions{
+					Donate: true,
+				},
+				Update: &kubebindv1alpha1.UpdateOptions{
+					Fields: []string{"spec"},
+				},
+			},
+			"The provider wants to create foo objects (apiVersion: \"v1\") on your cluster.\n" +
+				"The following fields of the objects will still be able to be changed by owner of this claim:\n" +
```
[the remark this answers is not in the document]
Comment (lhaendler) [reply pkg/kubectl/bind-apiservice/plugin/servicebindings_test.go:749]: "you" is also fine instead of "the user"

File: pkg/kubectl/bind-apiservice/plugin/servicebindings.go [hunk 12]
```diff
@@ -139,13 +141,156 @@ func (b *BindAPIServiceOptions) createAPIServiceBindings(ctx context.Context, co
...
+
+	if claim.Update.Fields != nil {
+		_, err = b.WriteString("The following fields of the objects will still be able to be changed by owner of this claim:\n")
+	}
+	if claim.Update.Preserving != nil {
+		_, err = b.WriteString("The following fields of the objects will be overwritten with their initial values, if they are modified:\n")
+	}
+
+	for _, s := range append(claim.Update.Fields, claim.Update.Preserving...) {
+		_, err = fmt.Fprintf(b, "	\"%s\"\n", s)
```
Comment (lhaendler) [inline pkg/kubectl/bind-apiservice/plugin/servicebindings.go:258]: Is this a tab? If yes, can you use `\t` instead?

File: pkg/kubectl/bind-apiservice/plugin/servicebindings_test.go [hunk 13]
```diff
@@ -0,0 +1,807 @@
...
+					Resource: "foo",
+				},
+				Version:  "v1",
+				Selector: kubebindv1alpha1.ResourceSelector{},
+				Required: true,
+				Create: &kubebindv1alpha1.CreateOptions{
+					Donate: true,
+				},
+			},
+			"The provider wants to create foo objects (apiVersion: \"v1\") on your cluster.\n" +
```
[the remark this answers is not in the document]
Comment (a9s-niklas-kempf) [reply pkg/kubectl/bind-apiservice/plugin/servicebindings_test.go]: Changed this in b9eac5dffe6fa930fd8ef4f2db5ead3c8cb1918f.

File: pkg/kubectl/bind-apiservice/plugin/servicebindings_test.go [hunk 14]
```diff
@@ -0,0 +1,807 @@
...
+				GroupResource: kubebindv1alpha1.GroupResource{
+					Group:    "",
+					Resource: "foo",
+				},
+				Version:  "v1",
+				Selector: kubebindv1alpha1.ResourceSelector{},
+				Required: true,
+				Adopt: true,
+			},
+			"The provider wants to have ownership of foo objects (apiVersion: \"v1\") on your cluster and assign the ownership to you.\n" +
```
[the remark this answers is not in the document]
Comment (a9s-niklas-kempf) [reply pkg/kubectl/bind-apiservice/plugin/servicebindings_test.go]: I wasn't sure about that, i removed what you suggested and also another case which cannot be reached if they are exclusive in b9eac5dffe6fa930fd8ef4f2db5ead3c8cb1918f.

File: pkg/kubectl/bind-apiservice/plugin/servicebindings.go (hunk 10 above)
Comment (a9s-niklas-kempf) [reply pkg/kubectl/bind-apiservice/plugin/servicebindings.go]: Looks like i missed to remove this after testing. I removed it in b9eac5dffe6fa930fd8ef4f2db5ead3c8cb1918f.

File: pkg/kubectl/bind-apiservice/plugin/servicebindings.go (hunk 12 above)
Comment (a9s-niklas-kempf) [reply pkg/kubectl/bind-apiservice/plugin/servicebindings.go]: I changed all "tabs" to `\t` in b9eac5dffe6fa930fd8ef4f2db5ead3c8cb1918f.

File: pkg/kubectl/bind-apiservice/plugin/servicebindings_test.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (a9s-niklas-kempf) [reply pkg/kubectl/bind-apiservice/plugin/servicebindings_test.go]: I changed to code and tests to use either "the user" or "the provider", depending on the `Donate` bool in b9eac5dffe6fa930fd8ef4f2db5ead3c8cb1918f.
